### PR TITLE
[MAINTENANCE] Refactor of `test_script_runner.py` to break-up test list

### PR DESCRIPTION
**Pull request by Shinnnyshinshin**
Changes proposed in this pull request:
- Small refactor of the `test_script_runner.py`'s `docs_test_matrix`, which is currently a (very) large list of dictionaries. I've tried breaking this list down into sub-lists based on type (cloud, local, docker) so that each sub-list can be switched on-off during development. 
- I was trying to add tests for the Migration guide, and this was an attempt to make things a little easier during development.
- Would love to hear ideas or suggestions too. 
- **NOTE**: The git `diff` seems to have some issues, but rest-assured that tests were moved around, and not altered in anyway :)

Previous Design Review notes:
- GREAT-504

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/en/latest/contributing/style_guide.html?highlight=style%20guide)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html?highlight=checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/en/latest/contributing/testing.html#contributing-testing-writing-unit-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

File: tests/integration/test_script_runner.py [hunk 1]
```diff
@@ -276,143 +206,248 @@ class BackendDependencies(enum.Enum):
...
+
+# populate docs_test_matrix with sub-lists
+docs_test_matrix += local_tests
+docs_test_matrix += dockerized_db_tests
+docs_test_matrix += cloud_snowflake_tests
+docs_test_matrix += cloud_gcp_tests
+docs_test_matrix += cloud_bigquery_tests
+docs_test_matrix += cloud_azure_tests
+docs_test_matrix += cloud_s3_tests
+docs_test_matrix += cloud_redshift_tests
```
Comment (Shinnnyshinshin) [inline tests/integration/test_script_runner.py:459]: This is what is enabled by this refactor

File: tests/integration/test_script_runner.py [hunk 2]
```diff
@@ -276,143 +206,248 @@ class BackendDependencies(enum.Enum):
...
+    # {
+    #     "user_flow_script": "tests/integration/docusaurus/connecting_to_your_data/cloud/azure/spark/inferred_and_runtime_python_example.py",
+    #     "data_context_dir": "tests/integration/fixtures/no_datasources/great_expectations",
+    # },
+]
+
+cloud_s3_tests = [
     {
-        "user_flow_script": "tests/integration/docusaurus/setup/configuring_data_docs/how_to_host_and_share_data_docs_on_gcs.py",
+        "user_flow_script": "tests/integration/docusaurus/connecting_to_your_data/cloud/s3/pandas/inferred_and_runtime_yaml_example.py",
```
Comment (Shinnnyshinshin) [inline tests/integration/test_script_runner.py:388]: git seems to have trouble resolving some `diffs` like these

**Review by NathanFarmer — APPROVED** [review]
LGTM! Once we have the tests grouped we can think about more efficient ways to run them.

File: tests/integration/test_script_runner.py [hunk 3]
```diff
@@ -23,15 +23,10 @@ class BackendDependencies(enum.Enum):
...
-        "user_flow_script": "tests/integration/docusaurus/deployment_patterns/gcp_deployment_patterns_file_gcs_yaml_configs.py",
-        "data_context_dir": "tests/integration/fixtures/no_datasources/great_expectations",
-    },
-    {
-        "user_flow_script": "tests/integration/docusaurus/deployment_patterns/gcp_deployment_patterns_file_bigquery_yaml_configs.py",
-        "data_context_dir": "tests/integration/fixtures/no_datasources/great_expectations",
-    },
+# to be populated by the smaller lists below
+docs_test_matrix = []
+
```
Comment (cdkini) [inline tests/integration/test_script_runner.py:28]: Could we add a blurb here about what this matrix is doing and what its contents look like?

Additionally, some information around the schema of the dictionary objects would be a huge help for newer folks to the team (what's required, what's optional, etc). I think a `TypedDict` would be ideal for this but it's 3.8-specific so we can't use it.

Perhaps a `namedtuple` could be handy to provide a lightweight schema but I don't think it's entirely necessary for a test runner like this.

**Review by cdkini — CHANGES_REQUESTED** [review]
All looks great but due to a number of newer folks getting into docs-under-test, I'd request we add a little bit to the documentation if you wouldn't mind.

A formal schema isn't entirely necessary but a docstring detailing the general structure would be a huge help for devs as we inch closer to 100% docs coverage.

**Review by cdkini — APPROVED** [review]

File: tests/integration/test_script_runner.py (hunk 3 above)
Comment (Shinnnyshinshin) [reply tests/integration/test_script_runner.py:28]: Thank you very much for this suggestion @cdkini, as we discussed I'll be including this change in subsequent PR

**Pull request merged**